### ISSUE TITLE
ZSH command capture doesn't work

### ISSUE THREAD

**Issue by xak2000** [dvorka/hstr#248]
> hh version "1.24" (2018-01-03T12:19:00)

For some reason in my ZSH installation (oh-my-zsh actually) default bindkey, generated by `hh -z` doesn't work as expected.

The generated line looks like this:

```
bindkey -s "\C-r" "\eqhh\n"     # bind hh to Ctrl-r (for Vi mode check doc)
```

It actually starts `hh` on pressing <kbd>CTRL-R</kbd>, but without passing already entered text in the prompt.

For example, if you type `$ apt`<kbd>CTRL-R</kbd> then `hh` will start, but without already entered `apt` text in the search line (like when you run `hh apt`). So, you forced to type `apt` again after pressing <kbd>CTRL-R</kbd> despite the fact you already entered it in your terminal before pressing <kbd>CTRL-R</kbd>.

I fixed this by defining a custom zle widget:

```
hh-start-search() {
    TMP=$BUFFER
    zle push-input
    BUFFER="hh $TMP"
    zle accept-line
}
zle -N hh-start-search
bindkey "\C-r" hh-start-search     # bind hh to Ctrl-r (for Vi mode check doc)
```

With this config in my `.zshrc`, <kbd>CTRL-R</kbd> works as expected (like in `bash`).
Maybe it is worth to generate something similar with `hh -z` instead of `bindkey -s "\C-r" "\eqhh\n"`?

**Comment by xak2000** [comment]
I explored the problem more and realize it is harder to do then I think.

### TLDR

The best zsh script that I could write is:

```
hh-start-search() {
    TMP=$BUFFER
    BUFFER=" hh -- $TMP"
    zle accept-line
}
```

Note the absense of `zle push-input` here, so if you cancel the history selection process, your command line will be empty. But this is compromise, because with presense of `zle push-input` line, the history selection doesn't work at all (it works, but runs wrong command, prefixed with text, which was entered before pressing <kbd>CTRL-R</kbd>, like `aptapt upgrade` if you enter `apt` before pressing hotkey).

--------------

The long version of my exploration is described [here](https://stackoverflow.com/questions/48711980/how-to-make-zsh-ctrl-r-start-hh-preserving-user-input).

Maybe it is possible to make both of this work:
1. Pass typed in text into `hh` and run selected command if I selected it. (_this works with the script above_)
2. Restore typed in text into command line prompt, if I decided to cancel history selection. (_doesn't work with script above_)

But I don't know how.

Another downside of this script - if you cancel the history selection, then ` hh -- apt` command somehow remains in the history (despite the space character at start of the line).

But, anyway, even with these downsides it is still better, then not working history selection (prefixing selected command with text in command line before hotkey was pressed).

**Comment by matt1003** [comment]
Is there an equivalent for this in bash?

**Comment by pquerner** [comment]
Can this be added to the documentation? The script works well with zsh as shell.

This is my current setup on MacOS High Sierra.
```
cat ~/.zshrc

# add this configuration to ~/.bashrc
export HH_CONFIG=hicolor         # get more colors
export HISTCONTROL=ignorespace   # leading space hides commands from history
export HISTFILESIZE=10000        # increase history file size (default is 500)
export HISTSIZE=${HISTFILESIZE}  # increase history size (default is 500)
export PROMPT_COMMAND="history -a; history -n; ${PROMPT_COMMAND}"   # mem/file sync

hh-start-search() {
    TMP=$BUFFER
    zle push-input
    BUFFER="hh $TMP"
    zle accept-line
}
zle -N hh-start-search
bindkey "\C-r" hh-start-search     # bind hh to Ctrl-r (for Vi mode check doc)
```

(bind and shopt are not available here)

**Comment by hoto** [comment]
I had the same problem when switching from bash to zsh.
In my `~/.zshrc` I changed the default `bindkey -s "\C-r" "\eqhh\n"` to `bindkey -s "\C-r" "\C-a hh -- \C-j"` and now it passes the previously typed input when pressing `Ctrl+r`.

**Comment by dvorka** [comment]
@hoto thank you for the suggestion - fix will be released w/ HSTR 2.1

**Comment by hoto** [comment]
Thanks @dvorka !  
Feels so good to contribute to the tool that I use every day!  
Wish you many github stars.
